scripts/feature_tests/upgrade/combined_tests/1cp_1w_bootDiskImageANDK8sControllers_clusterLevel_upgrade.sh

```diff
@@ -0,0 +1,190 @@
+  kubeadmconfigs.bootstrap.cluster.x-k8s.io -o json | jq '.spec.names.singular' | wc -l)
+upgraded_capm3_controller_count=$(kubectl api-resources | grep -c m3c2020)
+
+if [ "${upgraded_controllers_count}" -ne 2 ]; then
+  log_error "Failed to upgrade cluster-api and controlplane components"
+  log_test_result "1cp_1w_bootDiskImageANDK8sCotrollers_clusterLevel_upgrade.sh" "fail"
+  exit 1
+fi
+if [ "${upgraded_bootstrap_crd_count}" -ne 1 ]; then
+  log_error "Failed to upgrade control-plane-kubeadm components"
+  log_test_result "1cp_1w_bootDiskImageANDK8sCotrollers_clusterLevel_upgrade.sh" "fail"
+  exit 1
+fi
+
+if [ "${upgraded_capm3_controller_count}" -ne 1 ]; then
+  log_error "Failed to upgrade infrastructure components"
+  log_test_result "1cp_1w_bootDiskImageANDK8sCotrollers_clusterLevel_upgrade.sh" "fail"
+  exit 1
+fi
+
+sleep 30 # Wait for the controllers to be up and running
+
+health_controllers=$(kubectl get pods -A | grep -E "capm3-system|capi-kubeadm|metal3" |
+  grep -vc 'Running')
+if [ "${health_controllers}" -ne 0 ]; then
+  log_error "Some of the upgraded controlplane components are not healthy"
+  log_test_result "1cp_1w_bootDiskImageANDK8sCotrollers_clusterLevel_upgrade.sh" "fail"
+  exit 1
+fi
+
+# ----------- upgrade ironic image ------
+
+# Upgrade container images
+for container in ironic ironic-dnsmasq ironic-httpd mariadb; do
+  kubectl set image deployments metal3-ironic \
+    $container=quay.io/metal3-io/ironic:"${IRONIC_IMAGE_TAG}" -n "${NAMESPACE}"
+done
+kubectl set image deployments metal3-ironic \
+  ironic-inspector=quay.io/metal3-io/ironic-inspector:"${IRONIC_IMAGE_TAG}" -n "${NAMESPACE}"
+
+updated_ironic_image_count=$(kubectl get deployments metal3-ironic -n "${NAMESPACE}" -o json |
+  jq '.spec.template.spec.containers[].image' | grep -c "${IRONIC_IMAGE_TAG}")
+
+if [ "${updated_ironic_image_count}" -lt 5 ]; then
+  echo "All ironic images are not updated properly"
+  exit 1
+fi
+
+sleep 60 # wait until images are downloaded
+
+# Check if old and new ironic pods are running, wait until the old terminates
+ironic_pod_count=$(kubectl get pods -n "${NAMESPACE}" -o name | grep -c metal3-ironic)
+ironic_pod=$(kubectl get pods -n "${NAMESPACE}" -o name | grep metal3-ironic)
+
+if [ "${ironic_pod_count}" -gt 1 ]; then
+  for i in {1..300}; do
+	  sleep 15
+	  ironic_pod_count=$(kubectl get pods -n "${NAMESPACE}" -o name | grep -c metal3-ironic)
+	  if [ "${ironic_pod_count}" -eq 1 ]; then
+		  ironic_pod=$(kubectl get pods -n "${NAMESPACE}" -o name | grep metal3-ironic)
+		  break
+	  fi
+    if [[ "${i}" -ge 300 ]]; then
+      ironic_pod=$(kubectl get pods -n "${NAMESPACE}" -o name | grep metal3-ironic)
+      log_error " Ironic pods failed: ${ironic_pod}"
+      exit 1
+    fi
+  done
+else
+  echo "New ironic pod id: ${ironic_pod}"
+fi
+
+# Check that ironic containers are running
+for container in ironic ironic-inspector ironic-dnsmasq ironic-httpd mariadb; do
```

Comment:
Don't we have `ironic-endpoint-keepalived` container as well?

Reply:
If I remember that, all it does is attach an IP to an interface. It is not involved in interacting with nodes.